# Credit Card Fraud Detection using Machine Learning

# About Data:
This is a dataset containing credit card transactions with 31 features and a class label. The features represent various aspects of the transaction, and the class label indicates whether the transaction was fraudulent (class 1) or not (class 0).

The first feature is "Time", which represents the number of seconds elapsed between the transaction and the first transaction in the dataset. The next 28 features, V1 to V28, are anonymized variables resulting from a principal component analysis (PCA) transformation of the original features. They represent different aspects of the transaction, such as the amount, location, and type of transaction.

The second last feature is "Amount", which represents the transaction amount in USD. The last feature is the "Class" label, which indicates whether the transaction is fraudulent (class 1) or not (class 0).

Overall, this dataset is used to train machine learning models to detect fraudulent transactions in real-time. The features are used to train the model to learn patterns in the data, which can then be used to detect fraudulent transactions in future transactions.

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [2]:
credit_card_data = pd.read_csv('creditcard.csv')
credit_card_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [3]:
credit_card_data.sample()

Time       V1        V2        V3        V4        V5        V6  \
15633  27041.0 -0.39171 -0.327898  1.290511 -2.359108 -0.124416  0.465523   

             V7        V8        V9  ...       V21       V22       V23  \
15633 -0.156191  0.023381 -2.652046  ... -0.190081 -0.209513 -0.302544   

            V24       V25       V26       V27       V28  Amount  Class  
15633 -1.322573  0.462805 -0.060769  0.074085  0.056498   43.74      0  

[1 rows x 31 columns]

In [4]:
# dataset informations
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [5]:
# checking the number of missing values in each column
credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [6]:
# distribution of legit transactions & fraudulent transactions
credit_card_data['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

This Dataset is highly unblanced

0 --> Normal Transaction

1 --> fraudulent transaction

The first line of code creates a new dataframe called "legit" by selecting only the rows from the original "credit_card_data" dataframe where the "Class" label is equal to 0. In other words, it filters out all transactions labeled as fraudulent (Class == 1) and keeps only the legitimate transactions (Class == 0).

The second line of code creates a new dataframe called "fraud" by selecting only the rows from the original "credit_card_data" dataframe where the "Class" label is equal to 1. This filters out all legitimate transactions and keeps only the fraudulent transactions.

By separating the data into two dataframes, it becomes easier to analyze and compare the characteristics of legitimate and fraudulent transactions separately. This can be useful for identifying patterns or features that are more common in fraudulent transactions, which can then be used to develop models for fraud detection.

In [7]:
legit = credit_card_data[credit_card_data.Class==0]
fraud = credit_card_data[credit_card_data['Class']==1]

In [8]:
fraud['Class']

541       1
623       1
4920      1
6108      1
6329      1
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 492, dtype: int64

In [9]:
# statistical measures of the data
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

"count" indicates the total number of transactions in the dataset that have a valid "Amount" value (i.e., non-missing values).
"mean" is the average transaction amount for all the transactions in the dataset. In this case, the mean is 88.291022, indicating that the average transaction amount is around $88 USD.
"std" is the standard deviation of the transaction amounts. It is a measure of how spread out the transaction amounts are from the mean. In this case, the standard deviation is 250.105092, which indicates that the transaction amounts vary widely, with some transactions having very large amounts.
"min" is the smallest transaction amount in the dataset. In this case, the smallest transaction amount is 0.0, indicating that there may be some transactions with zero value or very low value.
"25%" is the first quartile of the transaction amounts. It indicates the value below which 25% of the transactions fall. In this case, the first quartile is 5.65, indicating that 25% of the transactions have a value less than or equal to $5.65 USD.
"50%" is the median transaction amount, which is the value that separates the lower half of the transactions from the upper half. In this case, the median is 22.0, indicating that half of the transactions have a value less than or equal to $22 USD.
"75%" is the third quartile of the transaction amounts. It indicates the value below which 75% of the transactions fall. In this case, the third quartile is 77.05, indicating that 75% of the transactions have a value less than or equal to $77.05 USD.
"max" is the largest transaction amount in the dataset. In this case, the largest transaction amount is 25,691.16, indicating that there may be some transactions with very large values.

In [10]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [11]:
# compare the values for both transactions
credit_card_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

Build a sample dataset containing similar distribution of normal transactions and Fraudulent Transactions

Number of Fraudulent Transactions --> 492

legit_sample = legit.sample(n=492) is a line of code that takes a random sample of 492 observations from the legit dataset. This is done to balance the number of observations in the legit and fraud datasets, which is necessary for training a machine learning model to predict fraud. Since the original dataset has a large number of legitimate transactions and a small number of fraudulent transactions, the model may be biased towards predicting that all transactions are legitimate. By creating a balanced dataset with an equal number of legitimate and fraudulent transactions, the model can be trained to better recognize the patterns that differentiate fraudulent transactions from legitimate ones

In [12]:
legit_sample = legit.sample(n=492)

In [13]:
new_df = pd.concat([legit_sample,fraud],axis=0)

In [14]:
new_df

Time        V1        V2        V3        V4        V5        V6  \
266088  162184.0  1.985257 -0.382365 -0.280022  0.494186 -0.721312 -0.528243   
149023   90464.0  1.951579 -0.272810 -1.549822  0.458287  0.470132 -0.300326   
148169   89506.0  0.142063  0.987541 -0.454654 -0.559020  0.952115 -0.802794   
191470  129268.0  1.454880 -1.136905 -2.226855  0.465189  0.041701 -0.868117   
70687    53979.0  1.408113 -0.626774  0.221465 -0.951568 -0.794761 -0.275469   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
266088 -0.574594 -0.077543  1.421329  ...  0.188784  0.823228  0.096345   
149023  0.166747 -0.302318  1.903295  ... -0.398406 -0.967664  0.167422   
148169  0.924056 -0.078269  0.024707  ... -0.353016 -0.890027  0.086538   
191470  0.774185 -0.356965  0.828902  ...  0.271675  0.103132 -0.250700   
70687  -0.756686  0.082900 -0.758601  ... -0.111780 -0.551403  0.030443   
...          ...       ...       ...  ...       ...       ...       ...   
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
266088 -0.074478 -0.064417 -0.207358  0.044080 -0.036908    9.99      0  
149023  0.039383 -0.140463  0.235328 -0.123197 -0.054288   92.29      0  
148169  0.461764 -0.419993  0.120486  0.219397  0.082660    1.29      0  
191470  0.561923  0.243882 -0.572658 -0.086792  0.000993  338.87      0  
70687  -0.560298  0.263457 -0.441789 -0.004327  0.002167   15.00      0  
...          ...       ...       ...       ...       ...     ...    ...  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[984 rows x 31 columns]

In [15]:
new_df['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

In [16]:
new_df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      92280.918699  0.070461 -0.009208  0.006534 -0.030870  0.012260   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.069850  0.149127 -0.156716  0.032329  ...  0.001803 -0.006146   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0      0.024097  0.008955  0.026066  0.022487 -0.010198 -0.004361 -0.003123   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0      102.486931  
1      122.211321  

[2 rows x 30 columns]

In [17]:
X = new_df.drop(columns='Class', axis=1)
Y = new_df['Class']

In [18]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=2)

# Model Training

Logistic Regression

In [19]:
model=LogisticRegression()

In [20]:
# training the Logistic Regression Model with Training Data
model.fit(X_train, Y_train)
# accuracy on training data
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)
print('Accuracy on Training data : ', training_data_accuracy)

Accuracy on Training data :  0.9174078780177891


In [21]:
# accuracy on test data
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)
print('Accuracy score on Test Data : ', test_data_accuracy)

Accuracy score on Test Data :  0.8883248730964467


In [22]:
X_test

Time         V1         V2         V3         V4         V5  \
247995  153761.0   1.146259   1.403458  -4.159148   2.660107  -0.323217   
278245  168114.0  -0.769170   0.751467  -0.519475  -0.298277   0.939478   
205983  136013.0  -2.030668  -0.556634   1.823039  -1.009048   0.590197   
237093  149089.0   1.010439  -2.342139  -1.062695  -0.255480  -1.542749   
15451    26833.0 -20.532751  12.373989 -23.009003   6.144821 -15.587296   
...          ...        ...        ...        ...        ...        ...   
108258   70828.0   0.196707   1.189757   0.704882   2.891388   0.045555   
12108    21046.0 -16.917468   9.669900 -23.736443  11.824990  -9.830548   
99506    67150.0  -1.824295   0.403327  -1.994122   2.756558  -3.139064   
74496    55614.0  -7.347955   2.397041  -7.572356   5.177819  -2.854838   
78684    57647.0   1.415236  -0.803360  -1.540675  -1.673538   1.651306   

              V6         V7         V8        V9  ...       V20       V21  \
247995 -1.836071  -1.623740   0.259562 -1.132044  ...  0.284831  0.564450   
278245 -1.139076   1.131914  -0.643690  0.784379  ... -0.106034  0.143352   
205983  2.445440  -0.830149   1.174764  1.171068  ... -0.305960  0.028729   
237093 -0.869002   0.066466  -0.273473  1.572977  ...  0.897036  0.498087   
15451  -4.384491 -15.939003  13.696416 -3.948455  ...  1.592754  1.754608   
...          ...        ...        ...       ...  ...       ...       ...   
108258  1.245730  -1.198714  -2.421616 -1.232089  ...  0.646616 -1.328132   
12108  -2.514829 -17.290657   1.820408 -6.264903  ...  0.996745 -2.336111   
99506   0.408185  -1.209045   1.095634 -1.447225  ...  1.503688  0.838760   
74496  -1.795239  -8.783235   0.437157 -3.740598  ...  0.073164 -0.175273   
78684   3.175625  -0.805772   0.692653 -0.899673  ...  0.241204  0.118095   

             V22       V23       V24       V25       V26       V27       V28  \
247995  0.445744 -0.141136 -0.265517  0.362260 -0.416062  0.507370  0.243744   
278245  1.065429 -0.153940 -0.090475 -0.440617 -0.244193 -0.331580 -0.109102   
205983  0.339005 -0.242841 -0.802692  0.230456  0.121284 -0.420461 -0.204697   
237093  0.376235 -0.285558  0.133299 -0.553592  0.646873 -0.142213  0.040607   
15451  -1.466115 -0.856779  0.125777  1.402587 -0.223755  1.574249  0.469201   
...          ...       ...       ...       ...       ...       ...       ...   
108258  0.189311 -0.005524 -0.814708  0.400924  0.286281  0.135215  0.257315   
12108   0.972755  1.241866 -1.051086  0.038009  0.672317  2.108471 -1.421243   
99506   0.341727  0.947506 -0.145493  0.049326  0.831065  0.332421  0.252713   
74496   0.543325 -0.547955 -0.503722 -0.310933 -0.163986  1.197895  0.378187   
78684   0.170406 -0.230289  1.058216  0.937143 -0.053590 -0.013110  0.003128   

        Amount  
247995   51.37  
278245   27.99  
205983   39.50  
237093  544.00  
15451    99.99  
...        ...  
108258    0.76  
12108     1.00  
99506   489.71  
74496     0.83  
78684    44.62  

[197 rows x 30 columns]

In [23]:
import pickle

In [24]:
with open('classifier.pkl','wb') as file:
    pickle.dump(model,file)